alpheios_nemo_ui/__init__.py

```diff
@@ -534,6 +535,147 @@ def new_level(self,subreference,prev):
+                    # the next reference is retained whole
+                    prevRange = str(refstart) + "-" + str(prevPassage)
+                    if index < len(reffs) - 1:
+                        nextRange = reffs[index+1]
+                    else:
+                        nextRange = None
+                    return prevRange, nextRange
+                else:
+                    # check to see if the custom reference is somewhere in the middle
+                    latest_start = self.latest_reff(refstart, substart)
+                    latest_end = self.latest_reff(refend,subend)
+                    if latest_start == substart and latest_end == refend:
+                        # custom reference is fully enclosed in the precalculated reference
+                        # previous reference is beginning of the precalculated reference
+                        # to the previous sibling and next reference is the next sibling to the
+                        # end of the precaculated reference
+                        return str(refstart) + "-" + str(prevPassage), str(nextPassage) + '-' + str(refend)
+        # if no matching range could be found, fall back to just the immeidate siblings as next and previous
+        return passage.siblingsId
+
+    def latest_reff(self, refA, refB):
+        """ Calculate which reff is the later one in the text
+            2.1.4 is later than 2.1
+            2.2 is later than 2.1.4
+            2.1.610 is later than 2.1.61
+        """
+        try:
+            lengthA = len(refA)
+            lengthB = len(refB)
+            if (lengthA != lengthB):
+            # one ref is longer than the other, need to check each part from the left
+                i = 0
+                while i < min(lengthA,lengthB) - 1:
+                    if float(refA.start.list[i]) > float(refB.start.list[i]):
+                        return refA
+                    elif float(refB.start.list[i]) > float(refA.start.list[i]):
+                        return refB
+                    i = i + 1
+
+                # if we get here then all were equal so we return the longest ref
+                if lengthA > lengthB:
+                    return refA
+                else:
+                    return refB
+            elif refA.parent == refB.parent:
+            # same length, same parent so we return the highest subreference
+                if float(refA.start.list[-1]) > float(refB.start.list[-1]):
+                    return refA
+                else :
+                    return refB
+            else:
+            # same length, different parent return the highest parent
+                if float(refA.parent.start.list[-1]) > float(refB.parent.start.list[-1]):
+                    return refA
+                else:
+                    return refB
+        except Exception as ex:
+            # if we have a reference with a string, we'll just quietly fail here
+            return None
```

Comment:
Is this for poems that have just letters in the reff? If they do have both letters and numbers maybe we can strip letters and then try to compare?

Reply:
some parts have letters and numbers and some have just letters. E.g. in Ovid Amores, we have a  poem identifier which is 'Ep' and references like 1.Ep.1

Other texts have a number and a letter, e.g. line 85a

But there is no way we can now for sure how to compare these, and they are more the exception than the rule, which is why I decided to just bail in this case. The worst case scenario is that we do line by line retrieval for those until we get to a range that doesn't have any letters in it.